index.bs

```diff
@@ -707,6 +711,37 @@ NOTE: The <a href="https://immersive-web.github.io/layers">WebXR layers module</
 
 </div>
 
+<div class="algorithm" data-algorithm="apply-frame-rate">
+
+When the user agent wants to <dfn>apply the frame rate</dfn> |rate| and an optional |promise| for an {{XRSession}} |session|, it MUST run the following steps:
+  1. [=Queue a task=] to perform the following steps:
+    1. Pass the new frame rate |rate| to the [=XR Compositor=].
+    1. The [=XR Compositor=] MAY use |rate| to calculate a new [=display frame rate=] and/or [=nominal frame rate=].
+    1. Let |oldrate| be the [=XRSession/internal framerate=].
+    1. Let |newrate| be the new [=nominal frame rate=].
+    1. [=Queue a task=] to perform the following steps:
+        1. If |newrate| is different from |oldrate| and |session|’s [=ended=] value is `false`, await until the {{XRSystem}}'s actions to update the [=nominal frame rate=] and [=display frame rate=] have taken effect.
+        1. Set |session|'s [=XRSession/internal framerate=] to |newrate|.
+        1. If |newrate| is different from |oldrate| and |session|’s [=ended=] value is `false`, fire an {{XRSessionEvent}} event named {{frameratechange!!event}} on |session|.
+        1. If |promise| is not `null`, [=/resolve=] |promise|.
+
+</div>
+
+<div class="algorithm" data-algorithm="update-frame-rate">
+
+The <dfn method for="XRSession">updateFrameRate(|rate|)</dfn> method passes the [=target frame rate=] |rate| to the {{XRSession}}.
```

Comment:
> I feel like there's a bug here because while this description states that the method is being passed the `target framerate` the algorithm doesn't actually set or store that anywhere, which makes it appear that this is setting the `internal framerate` directly with an opportunity for the compositor to veto, at which point any target frame rate expressed by the developer would effectively be lost.

Does the line `[=Apply the frame rate=] with |rate|, |promise| and |session|` not use this rate?
It's indeed true that the value that the developer set is indeed lost. I believe we discussed that they could store this themselves if they wanted to 

> Can we add an explicit `target framerate` slot to the session, like the existing `internal framerate`, that this algorithm sets to the value passed to `updateFrameRate()` as long as it's one of the supported framerates? That way it's made explicit that the `target frame rate` hint provided by developers doesn't simply vanish the next time the system updates it's internal framerate.

Yes, that sounds reasonable.